Adjust litterrobot tests and code to match guidelines

natekspencer opened the pull request:
<!--
  You are amazing! Thanks for contributing to our project!
  Please, DO NOT DELETE ANY TEXT from this template! (unless instructed).
-->
## Proposed change
<!--
  Describe the big picture of your changes here to communicate to the
  maintainers why we should accept this pull request. If it fixes a bug
  or resolves a feature request, be sure to link to that issue in the
  additional information section.
-->
Adjust the tests and code to match suggestions from previous PR reviews.

## Type of change
<!--
  What type of change does your PR introduce to Home Assistant?
  NOTE: Please, check only 1! box!
  If your PR requires multiple boxes to be checked, you'll most likely need to
  split it into multiple PRs. This makes things easier and faster to code review.
-->

- [x] Dependency upgrade
- [ ] Bugfix (non-breaking change which fixes an issue)
- [ ] New integration (thank you!)
- [x] New feature (which adds functionality to an existing integration)
- [ ] Breaking change (fix/feature causing existing functionality to break)
- [x] Code quality improvements to existing code or addition of tests

## Additional information
<!--
  Details are important, and help maintainers processing your PR.
  Please be sure to fill out additional details, if applicable.
-->

- This PR fixes or closes issue: fixes #
- This PR is related to issue: 
- Link to documentation pull request: 

## Checklist
<!--
  Put an `x` in the boxes that apply. You can also fill these out after
  creating the PR. If you're unsure about any of them, don't hesitate to ask.
  We're here to help! This is simply a reminder of what we are going to look
  for before merging your code.
-->

- [x] The code change is tested and works locally.
- [x] Local tests pass. **Your PR cannot be merged unless tests pass**
- [x] There is no commented out code in this PR.
- [x] I have followed the [development checklist][dev-checklist]
- [x] The code has been formatted using Black (`black --fast homeassistant tests`)
- [x] Tests have been added to verify that the new code works.

If user exposed functionality or configuration variables are added/changed:

- [ ] Documentation added/updated for [www.home-assistant.io][docs-repository]

If the code communicates with devices, web services, or third-party tools:

- [ ] The [manifest file][manifest-docs] has all fields filled out correctly.  
      Updated and included derived files by running: `python3 -m script.hassfest`.
- [ ] New or updated dependencies have been added to `requirements_all.txt`.  
      Updated by running `python3 -m script.gen_requirements_all`.
- [ ] Untested files have been added to `.coveragerc`.

The integration reached or maintains the following [Integration Quality Scale][quality-scale]:
<!--
  The Integration Quality Scale scores an integration on the code quality
  and user experience. Each level of the quality scale consists of a list
  of requirements. We highly recommend getting your integration scored!
-->

- [ ] No score or internal
- [ ] 🥈 Silver
- [x] 🥇 Gold
- [ ] 🏆 Platinum

<!--
  This project is very active and we have a high turnover of pull requests.

  Unfortunately, the number of incoming pull requests is higher than what our
  reviewers can review and merge so there is a long backlog of pull requests
  waiting for review. You can help here!
  
  By reviewing another pull request, you will help raise the code quality of
  that pull request and the final review will be faster. This way the general
  pace of pull request reviews will go up and your wait time will go down.
  
  When picking a pull request to review, try to choose one that hasn't yet
  been reviewed.

  Thanks for helping out!
-->

To help with the load of incoming pull requests:

- [ ] I have reviewed two other [open pull requests][prs] in this repository.

[prs]: https://github.com/home-assistant/core/pulls?q=is%3Aopen+is%3Apr+-author%3A%40me+-draft%3Atrue+-label%3Awaiting-for-upstream+sort%3Acreated-desc+review%3Anone

<!--
  Thank you for contributing <3

  Below, some useful links you could explore:
-->
[dev-checklist]: https://developers.home-assistant.io/docs/en/development_checklist.html
[manifest-docs]: https://developers.home-assistant.io/docs/en/creating_integration_manifest.html
[quality-scale]: https://developers.home-assistant.io/docs/en/next/integration_quality_scale_index.html
[docs-repository]: https://github.com/home-assistant/home-assistant.io

natekspencer committed:
Use SwitchEntity instead of ToggleEntity and adjust test patches as recommended

natekspencer committed:
Move async_create_entry out of try block in config_flow

natekspencer committed:
Patch pypi package instead of HA code

natekspencer commented:
@MartinHjelmare, I've made most of the changes you requested. A couple responses to your comments that I haven't addressed yet:

(https://github.com/home-assistant/core/pull/45886#discussion_r582446669): The entity_type that is passed in is controlled, so there isn't a risk of a trailing space at the moment.

(https://github.com/home-assistant/core/pull/46942#discussion_r582418784): Theoretically these can be separate sensors, but I don't know how relevant they are on their own. Sometimes it is interesting to see, just to know what is the basis for the "Waste Drawer" gauge calculation. Should they still be separated and/or removed?

(https://github.com/home-assistant/core/pull/46942#discussion_r582419445): Please see https://github.com/home-assistant/core/pull/46942#discussion_r580826534 for previous conversation on this and let me know how to proceed.

MartinHjelmare commented:
Times in state attributes should be UTC, there is no exception.

MartinHjelmare commented on homeassistant/components/litterrobot/hub.py:
```diff
@@ -101,10 +101,13 @@ def device_info(self):
     async def perform_action_and_refresh(self, action: MethodType, *args: Any):
         """Perform an action and initiates a refresh of the robot data after a few seconds."""
+
+        @callback
```
A callback can't be a coroutine function.
```suggestion
```

MartinHjelmare commented on tests/components/litterrobot/test_init.py:
```diff
@@ -1,20 +1,40 @@
+        "pylitterbot.Account.connect",
+        side_effect=exception,
+    ):
+        assert await hass.config_entries.async_setup(entry.entry_id) is False
```
We want to assert the state of the config entry. It should be `ENTRY_STATE_SETUP_RETRY` and `ENTRY_STATE_SETUP_ERROR` respectively.

MartinHjelmare commented:
If there's always an entity type passed in we should remove the other case.

MartinHjelmare commented:
What do these attributes measure?

As an example, a temperature measurement from a temperature sensor is relevant on its own, but a brightness measurement from a lamp isn't relevant without knowing if the lamp is on or off.

natekspencer committed:
Bump pylitterbot to 2021.2.6, fix tests, and implement other code review suggestions

natekspencer commented:
> > The entity_type that is passed in is controlled, so there isn't a risk of a trailing space at the moment.
> 
> If there's always an entity type passed in we should remove the other case.

Now I see what you are referring to. That was a relic of a previous version of entity_types in HACS that I hadn't removed. Thanks for the catch.

natekspencer commented:
> Times in state attributes should be UTC, there is no exception.

Is there a way to show only the time portion, whilst storing the absolute UTC time? The reason I ask is that the Litter-Robot API returns an epoch time for when the "sleep mode start time" is first set and is not updated unless the start time is manually changed. So if a user set that a month ago, there is now a month old "sleep mode start time", which makes zero sense to a user.

Sure, I could do some "magic" and update the date to be today's date. But then there is another question: do I always show the next start time, or do I show the previous start time while the unit is "sleeping" since it is an 8-hour window?

Also, as an end user of the robot, there are really only two pertinent data points associated with "sleep mode":
1. is sleep mode on or off
2. if on, what time is sleep mode set for

If there is a date component displayed, the user now has to decipher what it means to gain value out of it, which seems counter-intuitive to what home automation is all about.

Again, I look back on the input_datetime helper that only has a time component for how the history is tracked on it and it is precisely the functionality that makes sense for this data point. When I set a time on that helper, it doesn't care about a date and I only see state changes for things that make sense. There isn't a way to create an input_datetime helper through an integration is there? (I tried and got errors.) That would actually solve this altogether...

MartinHjelmare commented:
If the sleep mode measurements are supposed to show the start of the current sleep and the future end of the current sleep, I'd expect the time measurement to include the correct date besides the time.

Consider breaking out these measurements to sensors with device class timestamp. Timestamp sensors can be shown as relative time when using the Lovelace frontend entities card.

https://developers.home-assistant.io/docs/core/entity/sensor#available-device-classes

natekspencer committed:
Bump pylitterbot to 2021.2.8, remove sleep mode start/end time from vacuum, adjust and add sensors for sleep mode start/end time

natekspencer commented:
Alright, @MartinHjelmare, I moved those attributes to their own timestamp sensor with appropriate absolute UTC time and removed the other attributes on the waste drawer sensor per your guidelines!

MartinHjelmare commented on homeassistant/helpers/icon.py:
```diff
@@ -29,3 +29,14 @@ def icon_for_signal_level(signal_level: Optional[int] = None) -> str:
     return "mdi:signal-cellular-1"
+
+
+def icon_for_gauge_level(gauge_level: Optional[int] = None, offset: int = 0) -> str:
```
Please keep this helper in the integration for now.

MartinHjelmare commented on homeassistant/components/litterrobot/sensor.py:
```diff
@@ -36,19 +36,39 @@ def unit_of_measurement(self):
+    def state(self):
+        """Return the state."""
+        if self.robot.sleep_mode_active:
+            return super().state
```
Is this an an isoformatted datetime?
```suggestion
            return super().state
        return None
```

natekspencer committed:
Move icon helper back to Litter-Robot component and isoformat times on time sensors

MartinHjelmare reviewed (approved):
Thanks!

natekspencer commented:
> Thanks!

Thanks for all your help, @MartinHjelmare!

natekspencer commented:
@bdraco, part of the changes here resolve the issue mentioned in https://github.com/home-assistant/core/issues/47514, can we merge this pull request or should I create a different one only addressing/fixing that issue?

natekspencer commented on homeassistant/components/litterrobot/hub.py:
```diff
@@ -89,22 +89,21 @@ def unique_id(self):
 
     async def perform_action_and_refresh(self, action: MethodType, *args: Any):
         """Perform an action and initiates a refresh of the robot data after a few seconds."""
+
```
This is the specific section that resolves https://github.com/home-assistant/core/issues/47514

natekspencer commented on homeassistant/components/litterrobot/hub.py:
```diff
@@ -89,22 +89,21 @@ def unique_id(self):
+
+        async def async_call_later_callback(*_) -> None:
+            await self.hub.coordinator.async_request_refresh()
+
         await action(*args)
-        async_call_later(
-            self.hass, REFRESH_WAIT_TIME, self.hub.coordinator.async_request_refresh
-        )
+        async_call_later(self.hass, REFRESH_WAIT_TIME, async_call_later_callback)
```
This is the specific section that resolves https://github.com/home-assistant/core/issues/47514

MartinHjelmare replied:
I've tagged this PR for the next patch release.
